File: 42.接雨水.cpp
/*
 * @lc app=leetcode.cn id=42 lang=cpp
 *
 * [42] 接雨水
 */

// @lc code=start
class Solution {
public:
    int trap(vector<int>& height) {

        int n = height.size();
        if (n < 2) return 0;
        int ans = 0;

        stack<int> st;

        for (int i = 0; i < n; i ++ ) {

            while (st.size() && height[st.top()] < height[i]) {

                int first = st.top();
                st.pop();

                if (!st.size()) break;
                int second = st.top();

                int h = min(height[i], height[second]) - height[first];
                int w = i - second - 1;
                ans += (h * w);

            }
            st.push(i);
        }
        return ans;

    }
};
// @lc code=end

File: 42.接雨水.java
/*
 * @lc app=leetcode.cn id=42 lang=java
 *
 * [42] 接雨水
 */

// @lc code=start
class Solution {
    public int trap(int[] height) {

        int ans = 0;
        Stack<Integer> st = new Stack<>();

        for (int i = 0; i < height.length; i ++ ) {

            while (!st.isEmpty() && height[st.peek()] < height[i]) {
                int first = st.pop();
                if (st.isEmpty()) break;
                int second = st.peek();
                int h = Math.min(height[i], height[second]) - height[first];
                int w = i - second - 1;
                ans += (h * w);
            }
            st.add(i);
        }
        return ans;
    }
}
// @lc code=end

